presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSplitManager.java

```diff
@@ -0,0 +1,181 @@
+import static com.facebook.presto.hive.metastore.MetastoreUtil.extractPartitionValues;
+import static com.facebook.presto.hudi.HudiErrorCode.HUDI_FILESYSTEM_ERROR;
+import static com.facebook.presto.hudi.HudiErrorCode.HUDI_INVALID_METADATA;
+import static com.facebook.presto.hudi.HudiMetadata.fromDataColumns;
+import static com.facebook.presto.hudi.HudiMetadata.toMetastoreContext;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hudi.common.table.view.FileSystemViewManager.createInMemoryFileSystemViewWithTimeline;
+
+public class HudiSplitManager
+        implements ConnectorSplitManager
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final HudiTransactionManager hudiTransactionManager;
+    private final HudiPartitionManager hudiPartitionManager;
+
+    @Inject
+    public HudiSplitManager(
+            HdfsEnvironment hdfsEnvironment,
+            HudiTransactionManager hudiTransactionManager,
+            HudiPartitionManager hudiPartitionManager)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.hudiTransactionManager = requireNonNull(hudiTransactionManager, "hudiTransactionManager is null");
+        this.hudiPartitionManager = requireNonNull(hudiPartitionManager, "hudiPartitionManager is null");
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorTableLayoutHandle layoutHandle,
+            SplitSchedulingContext splitSchedulingContext)
+    {
+        ExtendedHiveMetastore metastore = ((HudiMetadata) hudiTransactionManager.get(transaction)).getMetastore();
+        HudiTableLayoutHandle layout = (HudiTableLayoutHandle) layoutHandle;
+        HudiTableHandle table = layout.getTable();
+
+        // Retrieve and prune partitions
+        List<String> partitions = hudiPartitionManager.getEffectivePartitions(session, metastore, table.getSchemaName(), table.getTableName(), layout.getTupleDomain());
+        if (partitions.isEmpty()) {
+            return new FixedSplitSource(ImmutableList.of());
+        }
+
+        // Load Hudi metadata
+        ExtendedFileSystem fs = getFileSystem(session, table);
+        HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().enable(false).build();
+        Configuration conf = fs.getConf();
+        HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder().setConf(conf).setBasePath(table.getPath()).build();
+        HoodieTimeline timeline = metaClient.getActiveTimeline().getCommitsTimeline().filterCompletedInstants();
+        HoodieLocalEngineContext engineContext = new HoodieLocalEngineContext(conf);
+        HoodieTableFileSystemView fsView = createInMemoryFileSystemViewWithTimeline(engineContext, metaClient, metadataConfig, timeline);
```

Comment:
Is this initialized for every partition or is it once per query?

Reply:
Per query.